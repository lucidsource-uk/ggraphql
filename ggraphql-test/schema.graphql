scalar DateTime @type(type: "java.util.Date")
scalar PaginationCursor @type(type: "uk.co.lucidsource.SimpleToken")

interface IdentifiedEntity {
    id: String
}

union ChangeItem = ChangeItemList | ChangeItemValue

type Candidate implements IdentifiedEntity {
    audits: [ChangeLog!]! @paginated @resolver(name: "CandidateResolver")
    firstname: String!
    id: String @filter(operators: [EQ])
    lastname: String! @filter(operators: [LIKE, EQ])
    status: CandidateStatus!
    tags: [String!] @filter(operators: [IN])
}

type CandidateStatusAggregation {
    interviewingCount: Int!
    notSeekingCount: Int!
    seekingCount: Int!
}

type ChangeItemList {
    leftValues: [String!]!
    property: String!
    rightValues: [String!]!
}

type ChangeItemValue {
    left: String
    property: String!
    right: String
}

type ChangeLog {
    changes: [ChangeItem!]!
    "ISO-8601"
    time: DateTime!
    type: ChangeLogType!
    user: User! @resolver(name: "UserResolver") @batchLoader
    otherUser: User @resolver(name: "UserResolver")
}

schema {
    query: QueryT,
    mutation: Mutation
}

"Mutation root"
type Mutation {
    persistCandidate(candidate: CandidateInput!): Candidate! @resolver(name: "CandidateResolver")
    deleteCandidate(candidateId: String!): SuccessResponse! @resolver(name: "CandidateResolver")
}

#"Query root"
type QueryT {
    aggregateByStatus: CandidateStatusAggregation!
    candidate(candidateId: String!): Candidate @resolver(name: "CandidateResolver")
    candidates: [Candidate!]! @paginated @filtered
    testNullArgument(input: CandidateInput, int: Int): String
    testNonNullArgument(input: CandidateInput!, int: Int!): String
    testSingleArgument(input: CandidateInput!): String
}

type SuccessResponse {
    success: Boolean!
}

type User {
    avatarUrl: String
    email: String!
    id: String!
}

enum CandidateStatus {
    INTERVIEWING
    NOT_SEEKING
    SEEKING
}

enum ChangeLogType {
    CREATED
    DELETED
    MODIFIED
}

input CandidateInput {
    id: String
    firstname: String!
    lastname: String!
    status: CandidateStatus!
    tags: [String]
}