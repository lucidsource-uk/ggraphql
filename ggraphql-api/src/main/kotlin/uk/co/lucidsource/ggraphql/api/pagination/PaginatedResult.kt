package uk.co.lucidsource.ggraphql.api.pagination

data class PaginatedResult<T, P>(
    val nodes: List<T>,
    val total: Int,
    val pageNumber: Int,
    val previousCursor: P? = null,
    val nextCursor: P? = null
)