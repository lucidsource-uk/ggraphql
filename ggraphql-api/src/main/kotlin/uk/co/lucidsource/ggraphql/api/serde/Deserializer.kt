package uk.co.lucidsource.ggraphql.api.serde

interface Deserializer {
    fun <O> deserialize(input: Any, clazz: Class<O>): O
}