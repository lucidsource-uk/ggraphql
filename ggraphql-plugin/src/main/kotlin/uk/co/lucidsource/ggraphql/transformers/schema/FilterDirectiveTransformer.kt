package uk.co.lucidsource.ggraphql.transformers.schema

import graphql.language.ArrayValue
import graphql.language.EnumValue
import graphql.language.FieldDefinition
import graphql.language.InputObjectTypeDefinition
import graphql.language.InputValueDefinition
import graphql.language.ListType
import graphql.language.NonNullType
import graphql.language.ObjectTypeDefinition
import graphql.language.TypeName
import uk.co.lucidsource.ggraphql.api.filtering.FilterGroupOperator
import uk.co.lucidsource.ggraphql.api.filtering.FilterOperator
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformer
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformerContext
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyFilterFieldNameForTypeAspect
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyFilterForTypeAspect
import uk.co.lucidsource.ggraphql.util.GraphQLTypeNameResolver
import uk.co.lucidsource.ggraphql.util.GraphQLTypeUtil

/**
 * Takes a GraphQL schema object types and expands the use of 'filter' annotations into
 * input types to build a filter query.
 */
class FilterDirectiveTransformer : SDLNodeTransformer {
    companion object {
        const val FILTER_DIRECTIVE = "filter"
        const val FILTER_DIRECTIVE_OPERATORS = "operators"
    }

    override fun transformObjectType(
        objectTypeDefinition: ObjectTypeDefinition, context: SDLNodeTransformerContext
    ): ObjectTypeDefinition {
        if (!shouldProcess(objectTypeDefinition)) {
            return objectTypeDefinition
        }

        // Generates filter criteria for each field in the form:
        // { operator: FieldType, otherOperator: FieldType }
        val filterFieldInputTypes = getFilterOperators(objectTypeDefinition)
            .map { fieldOperators ->
                val fieldName = fieldOperators.key.name

                fieldName to InputObjectTypeDefinition.newInputObjectDefinition()
                    .name(
                        GraphQLTypeNameResolver.getFilterFieldCriteriaTypeDefName(
                            objectTypeDefinition, fieldOperators.key
                        )
                    ).inputValueDefinitions(fieldOperators.value.map {
                        InputValueDefinition.newInputValueDefinition().name(it.fieldName).type(
                            if (it.isCompound) ListType(GraphQLTypeUtil.unwrapType(fieldOperators.key.type))
                            else GraphQLTypeUtil.unwrapType(fieldOperators.key.type)
                        ).build()
                    })
                    .applyFilterForTypeAspect(GraphQLTypeAspects.FilterAspectType.FIELD_CRITERIA)
                    .applyFilterFieldNameForTypeAspect(fieldName)
                    .build()
            }.toMap()

        // Generates filter criteria for all fields in the form
        // { field: FieldFilterCriteria, otherField: OtherFieldFilterCriteria }
        val filterCriteriaObjectTypeDef = InputObjectTypeDefinition.newInputObjectDefinition()
            .name(GraphQLTypeNameResolver.getFilterCriteriaTypeDefName(objectTypeDefinition))
            .inputValueDefinitions(filterFieldInputTypes.map {
                InputValueDefinition.newInputValueDefinition().name(it.key).type(TypeName(it.value.name)).build()
            })
            .applyFilterForTypeAspect(GraphQLTypeAspects.FilterAspectType.OBJECT_CRITERIA)
            .build()

        // Generates a filter in the form of { any: [FilterCriteria], not:[FilterCriteria] }
        val filterCriteriaObjectType =
            ListType(NonNullType(TypeName(GraphQLTypeNameResolver.getFilterCriteriaTypeDefName(objectTypeDefinition))))

        val filterObjectTypeDef = InputObjectTypeDefinition.newInputObjectDefinition()
            .name(GraphQLTypeNameResolver.getFilterTypeDefName(objectTypeDefinition)).inputValueDefinitions(
                FilterGroupOperator.entries.map {
                    InputValueDefinition.newInputValueDefinition()
                        .name(it.fieldName)
                        .type(filterCriteriaObjectType)
                        .build()
                }
            )
            .applyFilterForTypeAspect(GraphQLTypeAspects.FilterAspectType.EXPRESSION)
            .build()

        (filterFieldInputTypes.values + filterObjectTypeDef + filterCriteriaObjectTypeDef).forEach {
            context.newTypes.putIfAbsent(it.name, it)
        }

        return objectTypeDefinition
    }

    private fun getFilterOperators(typeDef: ObjectTypeDefinition): Map<FieldDefinition, List<FilterOperator>> {
        return typeDef.fieldDefinitions.filter { it.hasDirective(FILTER_DIRECTIVE) }
            .associate { field ->
                val filterDirective =
                    field.getDirectives(FILTER_DIRECTIVE)
                        .first()
                val arguments =
                    filterDirective.getArgument(FILTER_DIRECTIVE_OPERATORS).value as? ArrayValue
                val operators = (arguments?.values as? List<*>)?.filterIsInstance<EnumValue>()?.map {
                    FilterOperator.valueOf(it.name)
                }
                    ?: throw IllegalArgumentException("Unexpected type [${arguments.toString()}] for filter operators on field [${field.name}] of type [${typeDef.name}].")

                if (operators.isEmpty()) {
                    throw IllegalArgumentException("Filtered field [${field.name}] for type [${typeDef.name}] must supply valid filter operators.")
                }

                field to operators
            }
    }

    private fun shouldProcess(typeDef: ObjectTypeDefinition): Boolean {
        return typeDef.fieldDefinitions.any { it.hasDirective(FILTER_DIRECTIVE) }
    }
}