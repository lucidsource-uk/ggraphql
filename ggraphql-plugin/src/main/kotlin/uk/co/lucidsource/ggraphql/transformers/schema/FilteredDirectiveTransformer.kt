package uk.co.lucidsource.ggraphql.transformers.schema

import graphql.language.InputValueDefinition
import graphql.language.ObjectTypeDefinition
import graphql.language.TypeName
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformer
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformerContext
import uk.co.lucidsource.ggraphql.util.GraphQLTypeNameResolver

class FilteredDirectiveTransformer : SDLNodeTransformer {
    companion object {
        const val FILTERED_DIRECTIVE = "filtered"
    }

    override fun transformObjectType(
        objectTypeDefinition: ObjectTypeDefinition,
        context: SDLNodeTransformerContext
    ): ObjectTypeDefinition {
        if (!shouldProcess(objectTypeDefinition)) {
            return objectTypeDefinition
        }

        val originalFields = objectTypeDefinition
            .fieldDefinitions
            .filter { !it.hasDirective(FILTERED_DIRECTIVE) }

        val replacementFields = objectTypeDefinition
            .fieldDefinitions
            .filter { it.hasDirective(FILTERED_DIRECTIVE) }
            .map { field ->
                field.transform { fieldBuilder ->
                    fieldBuilder
                        .inputValueDefinition(
                            InputValueDefinition.newInputValueDefinition()
                                .name("where")
                                .type(TypeName(GraphQLTypeNameResolver.getFilterTypeDefName(field.type)))
                                .build()
                        )
                        .build()
                }
            }

        return objectTypeDefinition.transform { builder ->
            builder.fieldDefinitions(replacementFields + originalFields).build()
        }
    }

    private fun shouldProcess(typeDef: ObjectTypeDefinition): Boolean {
        return typeDef.fieldDefinitions.any { it.hasDirective(FILTERED_DIRECTIVE) }
    }
}