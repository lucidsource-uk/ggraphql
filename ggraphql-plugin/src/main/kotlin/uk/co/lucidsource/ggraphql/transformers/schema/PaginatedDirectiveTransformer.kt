package uk.co.lucidsource.ggraphql.transformers.schema

import graphql.language.FieldDefinition
import graphql.language.InputValueDefinition
import graphql.language.IntValue
import graphql.language.NonNullType
import graphql.language.ObjectTypeDefinition
import graphql.language.TypeName
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformer
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformerContext
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyExcludeFromCodeGenerationAspect
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyGeneratesResolverAspect
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyReturnsGenerateTypeParameterOf
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.isPaginatedAspectApplied
import uk.co.lucidsource.ggraphql.util.GraphQLTypeNameResolver
import uk.co.lucidsource.ggraphql.util.GraphQLTypeNameResolver.defaultResolverName
import java.math.BigInteger

/**
 * Takes object type fields annotated with 'paginated' and adds inputs for pagination, as well as changing the output type
 * to a paginated response.
 */
class PaginatedDirectiveTransformer : SDLNodeTransformer {
    companion object {
        const val DEFAULT_PAGE_SIZE = 10L
        const val NODES = "nodes"
        const val NEXT_CURSOR = "nextCursor"
        const val PREVIOUS_CURSOR = "previousCursor"
        const val CURSOR = "cursor"
        const val PAGE_NUMBER = "pageNumber"
        const val TOTAL = "total"
        const val PAGE_SIZE = "pageSize"
    }

    override fun transformObjectType(
        objectTypeDefinition: ObjectTypeDefinition,
        context: SDLNodeTransformerContext
    ): ObjectTypeDefinition {
        if (!shouldProcess(objectTypeDefinition)) {
            return objectTypeDefinition
        }

        val originalFields = objectTypeDefinition
            .fieldDefinitions
            .filter { !it.isPaginatedAspectApplied() }

        val replacementFields = objectTypeDefinition
            .fieldDefinitions
            .filter { it.isPaginatedAspectApplied() }
            .map { field ->
                field.transform { fieldBuilder ->
                    val paginationResultTypeDefName = GraphQLTypeNameResolver.getPaginationResultTypeDefName(field)

                    val paginationResultTypeDef = ObjectTypeDefinition.newObjectTypeDefinition()
                        .name(paginationResultTypeDefName)
                        .applyExcludeFromCodeGenerationAspect()
                        .fieldDefinitions(
                            listOf(
                                FieldDefinition.newFieldDefinition().name(NODES).type(field.type).build(),
                                FieldDefinition.newFieldDefinition().name(PAGE_NUMBER)
                                    .type(NonNullType(TypeName("Int"))).build(),
                                FieldDefinition.newFieldDefinition().name(TOTAL)
                                    .type(NonNullType(TypeName("Int"))).build(),
                                FieldDefinition.newFieldDefinition().name(NEXT_CURSOR)
                                    .type(TypeName("PaginationCursor")).build(),
                                FieldDefinition.newFieldDefinition().name(PREVIOUS_CURSOR)
                                    .type(TypeName("PaginationCursor")).build()
                            )
                        ).build()

                    context.newTypes.putIfAbsent(paginationResultTypeDefName, paginationResultTypeDef)

                    fieldBuilder
                        .inputValueDefinition(
                            InputValueDefinition.newInputValueDefinition().name(PAGE_SIZE)
                                .type(TypeName("Int")).defaultValue(IntValue(BigInteger.valueOf(DEFAULT_PAGE_SIZE)))
                                .build()
                        )
                        .inputValueDefinition(
                            InputValueDefinition.newInputValueDefinition().name(CURSOR)
                                .type(TypeName("String")).build()
                        )
                        .type(NonNullType(TypeName(paginationResultTypeDefName)))
                        .applyReturnsGenerateTypeParameterOf(field.type)
                        .applyGeneratesResolverAspect(objectTypeDefinition.defaultResolverName())
                        .build()
                }
            }

        return objectTypeDefinition.transform { builder ->
            builder.fieldDefinitions(originalFields + replacementFields).build()
        }
    }

    private fun shouldProcess(typeDef: ObjectTypeDefinition): Boolean {
        return typeDef.fieldDefinitions.any { it.isPaginatedAspectApplied() }
    }
}