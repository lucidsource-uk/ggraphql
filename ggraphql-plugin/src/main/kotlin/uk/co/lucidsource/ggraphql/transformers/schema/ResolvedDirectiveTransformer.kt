package uk.co.lucidsource.ggraphql.transformers.schema

import graphql.language.ObjectTypeDefinition
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformer
import uk.co.lucidsource.ggraphql.transformers.SDLNodeTransformerContext
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyExcludeFromCodeGenerationAspect
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.applyGeneratesResolverAspect
import uk.co.lucidsource.ggraphql.util.GraphQLTypeAspects.getResolverAspectResolverName
import uk.co.lucidsource.ggraphql.util.GraphQLTypeNameResolver.defaultResolverName

class ResolvedDirectiveTransformer : SDLNodeTransformer {
    override fun transformObjectType(
        objectTypeDefinition: ObjectTypeDefinition, context: SDLNodeTransformerContext
    ): ObjectTypeDefinition {
        if (objectTypeDefinition.name == context.queryRootName || objectTypeDefinition.name == context.mutationRootName) {
            return objectTypeDefinition.transform { objectBuilder ->
                val newFields =
                    objectTypeDefinition.fieldDefinitions
                        .filter { it.getResolverAspectResolverName() == null }
                        .map { field ->
                            field.transform { fieldBuilder ->
                                fieldBuilder.applyGeneratesResolverAspect(objectTypeDefinition.defaultResolverName())
                                    .build()
                            }
                        }

                objectBuilder.fieldDefinitions(newFields + objectTypeDefinition.fieldDefinitions
                    .filter {
                        it.getResolverAspectResolverName() != null
                    })

                objectBuilder
                    .applyExcludeFromCodeGenerationAspect()
                    .build()
            }
        }

        return objectTypeDefinition
    }
}