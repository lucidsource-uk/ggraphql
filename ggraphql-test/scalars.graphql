scalar DateTime @scalarType(type: "java.util.Date")
scalar PaginationCursor @scalarType(type: "uk.co.lucidsource.SimpleToken")