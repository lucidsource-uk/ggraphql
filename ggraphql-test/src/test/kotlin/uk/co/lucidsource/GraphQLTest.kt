package uk.co.lucidsource

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.junit5.SnapshotExtension
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.module.kotlin.registerKotlinModule
import graphql.GraphQL
import graphql.Scalars
import graphql.execution.AsyncExecutionStrategy
import graphql.scalars.ExtendedScalars
import graphql.schema.GraphQLCodeRegistry.newCodeRegistry
import graphql.schema.GraphQLScalarType
import graphql.schema.idl.RuntimeWiring
import graphql.schema.idl.SchemaGenerator
import graphql.schema.idl.SchemaParser
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import uk.co.lucidsource.generated.models.Candidate
import uk.co.lucidsource.generated.models.CandidateFilter
import uk.co.lucidsource.generated.models.CandidateInput
import uk.co.lucidsource.generated.resolvers.CandidateResolver
import uk.co.lucidsource.generated.models.CandidateStatus
import uk.co.lucidsource.generated.models.CandidateStatusAggregation
import uk.co.lucidsource.generated.models.ChangeItemList
import uk.co.lucidsource.generated.models.ChangeItemValue
import uk.co.lucidsource.generated.models.ChangeLog
import uk.co.lucidsource.generated.models.ChangeLogType
import uk.co.lucidsource.generated.resolvers.GraphQLCodeDataFetcherRegistry
import uk.co.lucidsource.generated.resolvers.QueryTResolver
import uk.co.lucidsource.generated.models.SuccessResponse
import uk.co.lucidsource.generated.models.User
import uk.co.lucidsource.generated.resolvers.UserResolver
import uk.co.lucidsource.generated.wiring.TypeResolverWiring
import uk.co.lucidsource.ggraphql.api.pagination.PaginatedResult
import uk.co.lucidsource.ggraphql.api.serde.Deserializer
import java.io.File
import java.util.Date

@ExtendWith(value = [SnapshotExtension::class])
class GraphQLTest {
    lateinit var expect: Expect

    companion object {
        val CANDIDATES: List<Candidate> = listOf(
            Candidate(
                id = "1",
                firstname = "John",
                lastname = "Doe",
                tags = listOf("react"),
                status = CandidateStatus.NOT_SEEKING
            ),
            Candidate(
                id = "2",
                firstname = "Jane",
                lastname = "Doe",
                tags = null,
                status = CandidateStatus.INTERVIEWING
            )
        )

        val CHANGE_LOGS: List<ChangeLog> = listOf(
            ChangeLog(
                changes = listOf(
                    ChangeItemList(leftValues = listOf(), property = "tags", rightValues = listOf()),
                    ChangeItemValue(left = "john", property = "firstname", right = "doe")
                ),
                time = Date(),
                type = ChangeLogType.CREATED
            )
        )
    }

    class MockCandidateResolver(
        val candidates: MutableList<Candidate>,
        val changesLogs: MutableList<ChangeLog>
    ) : CandidateResolver {
        override fun audits(
            pageSize: Int?,
            cursor: SimpleToken?,
            candidate: Candidate
        ): PaginatedResult<ChangeLog, SimpleToken> {
            return PaginatedResult(
                nodes = changesLogs,
                pageNumber = 1,
                total = changesLogs.size,
                nextCursor = SimpleToken("sdsd")
            )
        }

        override fun persistCandidate(candidate: CandidateInput): Candidate {
            val newCandidate = Candidate(
                id = "99",
                firstname = candidate.firstname,
                lastname = candidate.lastname,
                tags = candidate.tags?.filterNotNull(),
                status = candidate.status
            )
            candidates.add(newCandidate)
            return newCandidate
        }

        override fun deleteCandidate(candidateId: String): SuccessResponse {
            val candidate = candidates.firstOrNull { it.id == candidateId }
            if (candidate != null) {
                candidates.remove(candidate)
            }
            return SuccessResponse(candidate != null)
        }

        override fun candidate(candidateId: String): Candidate? {
            return candidates.firstOrNull { it.id == candidateId }
        }
    }

    class MockQueryResolver(
        val candidates: MutableList<Candidate>
    ) : QueryTResolver {
        override fun aggregateByStatus(): CandidateStatusAggregation {
            return CandidateStatusAggregation(
                interviewingCount = candidates.count { it.status == CandidateStatus.INTERVIEWING },
                seekingCount = candidates.count { it.status == CandidateStatus.SEEKING },
                notSeekingCount = candidates.count { it.status == CandidateStatus.NOT_SEEKING }
            )
        }

        override fun testNullArgument(input: CandidateInput?, int: Int?): String? {
            return "worked"
        }

        override fun testNonNullArgument(input: CandidateInput, int: Int): String? {
            return "worked"
        }

        override fun testSingleArgument(input: CandidateInput): String? {
            return "worked"
        }

        override fun candidates(
            where: CandidateFilter?,
            pageSize: Int?,
            cursor: SimpleToken?
        ): PaginatedResult<Candidate, SimpleToken> {
            return PaginatedResult(nodes = candidates, pageNumber = 1, total = candidates.size)
        }
    }

    class MockUserResolver : UserResolver {
        override fun user(changeLog: ChangeLog): User? {
            return User(id = "1", email = "test@emai.com", avatarUrl = null)
        }
    }

    fun buildSchema(candidates: MutableList<Candidate>, changeLogs: MutableList<ChangeLog>): GraphQL {
        val typeDefinitionRegistry = SchemaParser()
            .parse(File("src/test/resources/schema.graphql"))

        val deserializer = object : Deserializer {
            override fun <O> deserialize(input: Any, clazz: Class<O>): O {
                return ObjectMapper().registerKotlinModule().convertValue(input, clazz)
            }
        }

        val wiring = RuntimeWiring.newRuntimeWiring()
            .codeRegistry(
                GraphQLCodeDataFetcherRegistry(
                    candidateResolver = MockCandidateResolver(candidates, changeLogs),
                    queryTResolver = MockQueryResolver(candidates),
                    userResolver = MockUserResolver(),
                    deserializer = deserializer
                ).registerResolvers(newCodeRegistry()).build()
            )

        wiring.scalar(ExtendedScalars.DateTime)
        wiring.scalar(
            GraphQLScalarType.newScalar(Scalars.GraphQLString)
                .name("PaginationCursor")
                .coercing(PaginationCursorCoercing())
                .build()
        )

        TypeResolverWiring.wireTypeResolvers(wiring)

        val schema = SchemaGenerator()
            .makeExecutableSchema(typeDefinitionRegistry, wiring.build())

        return GraphQL.newGraphQL(schema)
            .queryExecutionStrategy(AsyncExecutionStrategy())
            .build()
    }

    @Test
    fun testQueryUnionResults() {
        val graphQL = buildSchema(CANDIDATES.toMutableList(), CHANGE_LOGS.toMutableList())

        val result =
            graphQL.execute("""{ candidate(candidateId: "1") { audits { nextCursor, nodes { changes { ... on ChangeItemValue { left, right, property } ... on ChangeItemList { leftValues, property, rightValues } } } } } }""")

        expect.toMatchSnapshot(result)
    }

    @Test
    fun testQueryPaginatedResults() {
        val graphQL = buildSchema(CANDIDATES.toMutableList(), CHANGE_LOGS.toMutableList())

        val result =
            graphQL.execute("""{ candidates(cursor: "test") { nodes { id, firstname, lastname }, total, pageNumber } }""")

        expect.toMatchSnapshot(result)
    }

    @Test
    fun testMutation() {
        val graphQL = buildSchema(CANDIDATES.toMutableList(), CHANGE_LOGS.toMutableList())

        val mutationResult =
            graphQL.execute("""mutation { persistCandidate(candidate: {firstname: "new", lastname: "new", status: SEEKING} ) { id, firstname, lastname } }""")

        val queryResult =
            graphQL.execute("""{ candidates { nodes { id, firstname, lastname }, total, pageNumber } }""")

        expect.scenario("mutation").toMatchSnapshot(mutationResult)
        expect.scenario("query").toMatchSnapshot(queryResult)
    }

    @Test
    fun testNullArgumentWithNull() {
        val graphQL = buildSchema(CANDIDATES.toMutableList(), CHANGE_LOGS.toMutableList())

        val queryResult =
            graphQL.execute("""{ testNullArgument }""")

        expect.scenario("query").toMatchSnapshot(queryResult)
    }

    @Test
    fun testNonNullArgumentWithNull() {
        val graphQL = buildSchema(CANDIDATES.toMutableList(), CHANGE_LOGS.toMutableList())

        val queryResult =
            graphQL.execute("""{ testNonNullArgument }""")

        expect.scenario("query").toMatchSnapshot(queryResult)
    }

    @Test
    fun testNonNullArgumentWithNonNull() {
        val graphQL = buildSchema(CANDIDATES.toMutableList(), CHANGE_LOGS.toMutableList())

        val queryResult =
            graphQL.execute("""{ testNonNullArgument(input: {firstname: "new", lastname: "new", status: SEEKING}, int: 5) }""")

        expect.scenario("query").toMatchSnapshot(queryResult)
    }
}