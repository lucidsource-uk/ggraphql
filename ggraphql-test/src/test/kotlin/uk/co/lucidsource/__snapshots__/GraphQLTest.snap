uk.co.lucidsource.GraphQLTest.testMutation[mutation]=[
ExecutionResultImpl{errors=[], data={persistCandidate={id=99, firstname=new, lastname=new}}, dataPresent=true, extensions=null}
]


uk.co.lucidsource.GraphQLTest.testMutation[query]=[
ExecutionResultImpl{errors=[], data={candidates={nodes=[{id=1, firstname=John, lastname=Doe}, {id=2, firstname=Jane, lastname=Doe}, {id=99, firstname=new, lastname=new}], total=3, pageNumber=1}}, dataPresent=true, extensions=null}
]


uk.co.lucidsource.GraphQLTest.testNonNullArgumentWithNonNull[query]=[
ExecutionResultImpl{errors=[], data={testNonNullArgument=worked}, dataPresent=true, extensions=null}
]


uk.co.lucidsource.GraphQLTest.testNonNullArgumentWithNull[query]=[
ExecutionResultImpl{errors=[ValidationError{validationErrorType=MissingFieldArgument, queryPath=[testNonNullArgument], message=Validation error (MissingFieldArgument@[testNonNullArgument]) : Missing field argument 'input', locations=[SourceLocation{line=1, column=3}], description='Validation error (MissingFieldArgument@[testNonNullArgument]) : Missing field argument 'input'', extensions=[]}, ValidationError{validationErrorType=MissingFieldArgument, queryPath=[testNonNullArgument], message=Validation error (MissingFieldArgument@[testNonNullArgument]) : Missing field argument 'int', locations=[SourceLocation{line=1, column=3}], description='Validation error (MissingFieldArgument@[testNonNullArgument]) : Missing field argument 'int'', extensions=[]}], data=null, dataPresent=false, extensions=null}
]


uk.co.lucidsource.GraphQLTest.testNullArgumentWithNull[query]=[
ExecutionResultImpl{errors=[], data={testNullArgument=worked}, dataPresent=true, extensions=null}
]


uk.co.lucidsource.GraphQLTest.testQueryPaginatedResults=[
ExecutionResultImpl{errors=[], data={candidates={nodes=[{id=1, firstname=John, lastname=Doe}, {id=2, firstname=Jane, lastname=Doe}], total=2, pageNumber=1}}, dataPresent=true, extensions=null}
]


uk.co.lucidsource.GraphQLTest.testQueryUnionResults=[
ExecutionResultImpl{errors=[], data={candidate={audits={nextCursor=SimpleToken(token=sdsd), nodes=[{changes=[{leftValues=[], property=tags, rightValues=[]}, {left=john, right=doe, property=firstname}]}]}}}, dataPresent=true, extensions=null}
]